### PR TITLE
Implement scatter for struct columns

### LINKED ISSUE #7638: [BUG] cudf::partition causing data corruption on structs
**Describe the bug**
We recently switched over to try and use cudf::partition (0.19) as a way to partition shuffle data in some cases for Spark. Instantly a test started to fail that partitioning a column with a struct in it.

**Steps/Code to reproduce bug**
```diff
diff --git a/cpp/tests/partitioning/partition_test.cpp b/cpp/tests/partitioning/partition_test.cpp
index a6838112a5..e15d57fc67 100644
--- a/cpp/tests/partitioning/partition_test.cpp
+++ b/cpp/tests/partitioning/partition_test.cpp
@@ -141,6 +141,35 @@ TYPED_TEST(PartitionTest, Identity)
   run_partition_test(table_to_partition, map, 6, table_to_partition, expected_offsets);
 }
 
+TYPED_TEST(PartitionTest, Struct)
+{
+  using value_type = cudf::test::GetType<TypeParam, 0>;
+  using map_type   = cudf::test::GetType<TypeParam, 1>;
+
+  fixed_width_column_wrapper<value_type, int32_t> A({1, 2}, {0, 1});
+  auto struct_col = cudf::test::structs_column_wrapper({A}, {0, 1}).release();
+  auto table_to_partition = cudf::table_view{{*struct_col}};
+
+  fixed_width_column_wrapper<map_type> map{9, 2};
+
+  fixed_width_column_wrapper<value_type, int32_t> A_expected({2, 1}, {1, 0});
+  auto struct_expected = cudf::test::structs_column_wrapper({A_expected}, {1, 0}).release();
+  auto expected = cudf::table_view{{*struct_expected}};
+
+  std::vector<cudf::size_type> expected_offsets{0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2};
+
+  // This does not work because we cannot sort a struct right now...
+  // run_partition_test(table_to_partition, map, 12, expected, expected_offsets);
+  // But there is no ambiguity in the ordering so I'll just copy it all here for now.
+  auto num_partitions = 12;
+  auto result = cudf::partition(table_to_partition, map, num_partitions);
+  auto const& actual_partitioned_table = result.first;
+  auto const& actual_offsets           = result.second;
+  EXPECT_EQ(actual_offsets, expected_offsets);
+
+  CUDF_TEST_EXPECT_TABLES_EQUAL(expected, *actual_partitioned_table);
+}
+
 TYPED_TEST(PartitionTest, Reverse)
 {
   using value_type = cudf::test::GetType<TypeParam, 0>;
```
This test definitely fails and I think it is reproducing the error.  Essentially what I see is that the child of the struct is not being changed despite the parent column being reordered. I think this is a scatter issue, but I am not 100% sure.

**Expected behavior**
The child of the struct should be reordered just like the struct was. 

**Additional context**
Please verify that lists are working correctly too.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR implements scattering for struct columns. It also fixes #7638 when calling `cudf::partitioning` on struct data due to wrongly scattering data inside the partitioning kernels.

### CODE REVIEW

**Comment by ttnghia** [comment]
Currently I'm working on unit tests for struct scattering, which should be done shortly.

**Review by revans2 — APPROVED** [review]
From the java side this fixes the issues I was seeing before.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

File: cpp/include/cudf/detail/scatter.cuh [hunk 1]
```diff
@@ -62,11 +64,16 @@ auto scatter_to_gather(MapIterator scatter_map_begin,
...
+  // entries when calling the gather_bitmask() which applies a pass-through whenever it finds a
   // value outside the range of the target column.
-  // We'll use the gather_rows value for this since it should always be outside the valid range.
-  auto gather_map = rmm::device_vector<size_type>(gather_rows, gather_rows);
+  // We'll use the `numeric_limits::lowest()` value for this since it should always be outside the
+  // valid range.
+  auto gather_map = rmm::device_uvector<size_type>(gather_rows, stream);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             gather_map.begin(),
+                             gather_map.end(),
```
Comment (codereport) [inline cpp/include/cudf/detail/scatter.cuh:75]:
```suggestion
                             gather_map.cbegin(),
                             gather_map.cend(),
```

**Review by codereport — CHANGES_REQUESTED** [review]
looks great overal, couple small things

File: cpp/include/cudf/detail/scatter.cuh [hunk 2]
```diff
@@ -299,11 +420,16 @@ std::unique_ptr<table> scatter(
...
-  auto gather_map = scatter_to_gather(
-    updated_scatter_map_begin, updated_scatter_map_end, target.num_rows(), stream);
-
-  gather_bitmask(source, gather_map.begin(), result, gather_bitmask_op::PASSTHROUGH, stream, mr);
-
+  // We still need to call `gather_bitmask` even when the source columns are not nullable,
+  // as if the target has null_mask, that null_mask needs to be updated after scattering
+  auto const nullable =
+    std::any_of(source.begin(), source.end(), [](auto const& col) { return col.nullable(); }) or
+    std::any_of(target.begin(), target.end(), [](auto const& col) { return col.nullable(); });
```
Comment (codereport) [inline cpp/include/cudf/detail/scatter.cuh:427]:
```suggestion
    std::any_of(source.cbegin(), source.cend(), [](auto const& col) { return col.nullable(); }) or
    std::any_of(target.cbegin(), target.cend(), [](auto const& col) { return col.nullable(); });
```

**Review by codereport — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/detail/scatter.cuh (hunk 1 above)
Comment (ttnghia) [reply cpp/include/cudf/detail/scatter.cuh:75]: Hey, we are writing to those iterators.

File: cpp/include/cudf/detail/scatter.cuh (hunk 2 above)
Comment (ttnghia) [reply cpp/include/cudf/detail/scatter.cuh:427]: Unfortunately `source` and `target` are `table_view`, which doesn't have const iterators.

File: cpp/include/cudf/detail/scatter.cuh (hunk 1 above)
Comment (codereport) [reply cpp/include/cudf/detail/scatter.cuh:75]: :facepalm:

**Review by codereport — APPROVED** [review]
LGTM then :+1:

**Review by kkraus14 — APPROVED** [review]
CMake lgtm

**Review by jrhemstad — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**